File: leetcode/542.01-matrix.py
# @leet start
class Solution:
    def updateMatrix(self, mat: list[list[int]]) -> list[list[int]]:
        NOT_VISITED = -1
        rows = len(mat)
        cols = len(mat[0])

        layer: list[tuple[int, int]] = []
        res = [[NOT_VISITED for _ in range(cols)] for _ in range(rows)]

        # Put zero cells in a list
        for row in range(rows):
            for col in range(cols):
                if mat[row][col] == 0:
                    layer.append((row, col))
                    res[row][col] = 0

        # Do multi-source BFS
        dist = 0

        while layer:
            next_layer: list[tuple[int, int]] = []
            dist += 1

            for row, col in layer:
                for delta_row, delta_col in ((0, 1), (0, -1), (1, 0), (-1, 0)):
                    adj_row = row + delta_row
                    adj_col = col + delta_col

                    if (
                        0 <= adj_row < rows
                        and 0 <= adj_col < cols
                        and res[adj_row][adj_col] == NOT_VISITED
                    ):
                        next_layer.append((adj_row, adj_col))
                        res[adj_row][adj_col] = dist

            layer = next_layer

        return res


# @leet end

File: leetcode/542.01-matrix.ts
// @leet start
function updateMatrix(mat: number[][]): number[][] {
  // Go left to right top to bottom finding the heights, then update
  // again going the other direction
  const rows = mat.length;
  const cols = mat[0].length;

  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      if (mat[row][col] !== 0) {
        mat[row][col] =
          1 +
          Math.min(
            row > 0 ? mat[row - 1][col] : Infinity,
            col > 0 ? mat[row][col - 1] : Infinity,
          );
      }
    }
  }

  for (let row = rows - 1; row >= 0; row--) {
    for (let col = cols - 1; col >= 0; col--) {
      if (mat[row][col] !== 0) {
        mat[row][col] = Math.min(
          mat[row][col],
          1 +
            Math.min(
              row < rows - 1 ? mat[row + 1][col] : Infinity,
              col < cols - 1 ? mat[row][col + 1] : Infinity,
            ),
        );
      }
    }
  }

  return mat;
}
// @leet end

